File: src/main/java/com/indiepost/repository/ContributorRepository.java
package com.indiepost.repository;

import com.indiepost.enums.Types;
import com.indiepost.model.Contributor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

public interface ContributorRepository {

    Long save(Contributor contributor);

    void delete(Contributor contributor);

    void deleteById(Long id);

    Optional<Contributor> findById(Long id);

    Long count();

    Page<Contributor> findAll(Pageable pageable);

    Page<Contributor> findAllByContributorType(Types.ContributorType contributorType, Pageable pageable);

    Page<Contributor> findAllByFullName(String fullName, Pageable pageable);

    Page<Contributor> findByIdIn(List<Long> ids, Pageable pageable);

    Page<Contributor> findByFullNameIn(List<String> contributorNames, Pageable pageable);

    Long countAllByContributorType(Types.ContributorType contributorType);

}

File: src/main/kotlin/com/indiepost/repository/ContributorRepository.kt
package com.indiepost.repository

import com.indiepost.enums.Types
import com.indiepost.model.Contributor
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable

interface ContributorRepository {

    fun save(contributor: Contributor): Long?

    fun delete(contributor: Contributor)

    fun deleteById(id: Long)

    fun findById(id: Long): Contributor?

    fun count(): Long

    fun findAll(pageable: Pageable): Page<Contributor>

    fun findAllByContributorType(contributorType: Types.ContributorType, pageable: Pageable): Page<Contributor>

    fun findAllByFullName(fullName: String, pageable: Pageable): Page<Contributor>

    fun findByIdIn(ids: List<Long>, pageable: Pageable): Page<Contributor>

    fun findByFullNameIn(contributorNames: List<String>, pageable: Pageable): Page<Contributor>

    fun countAllByContributorType(contributorType: Types.ContributorType): Long

}
